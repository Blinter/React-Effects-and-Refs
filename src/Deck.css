/* Deck.css */
.card-container {
  position: absolute;
  top: 50%;
  left: 50%;
  width: 400px;
  height: 500px;
  margin: 20px auto;
  margin: -250px 0 0 -250px;
  perspective: 1000px;
}

.card-flip {
  position: absolute;
  width: 100%;
  height: 100%;
  transform-style: preserve-3d;
  transition: transform 0.5s;
}

.card-flip.flipped {
  transform: rotateY(180deg);
}

.card-face {
  position: absolute;
  width: 100%;
  height: 100%;
  backface-visibility: hidden;
  border-radius: 10px;
  overflow: hidden;
}

.card-back {
  background-color: #007bff;
  color: white;
  display: flex;
  justify-content: center;
  align-items: center;
  font-size: 24px;
  transform: rotateY(180deg);
}

.card-image {
  width: 100%;
  height: 100%;
  object-fit: fill;
}