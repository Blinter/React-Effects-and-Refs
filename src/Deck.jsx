import React, { useState, useEffect } from 'react';
import axios from 'axios';
import './Deck.css';
import { v4 as uuidv4 } from 'uuid';
import DeckControls from './DeckControls';
import Card from './Card';
import useInterval from './useInterval';

function Deck() {
    const [deck, setDeck] = useState([]);
    const [deckId, setDeckId] = useState(null);
    const [cardsLeft, setCardsLeft] = useState(-1);
    const [isLoading, setIsLoading] = useState(true);
    const [rotations, setRotations] = useState([]);
    const [autoDrawEnabled, setAutoDrawEnabled] = useState(false);

    const toggleAutoDraw = () => {
        setAutoDrawEnabled(prevEnable => !prevEnable);
    };

    useInterval(
        () => {
            if (cardsLeft > 0)
                if (deckId == null)
                    showCard();
                else
                    addCard();
            else
                setAutoDrawEnabled(false);
        },
        autoDrawEnabled ? 500 : null
    );

    const generateRotation = () => {
        return `rotate(${Math.floor(Math.random() * 360)}deg)`;
    };

    const addCard = async () => {
        if (deckId === null) {
            console.error("deckId is null!");
            return;
        }

        try {
            const response = await axios.get(`https://deckofcardsapi.com/api/deck/${deckId}/draw/?count=1`);
            if (!response ||
                !response.data) {
                throw new Error('Invalid API response');
            }
            if (response == null)
                return;
            const deckData = response.data;
            if (!deckData ||
                !deckData.cards ||
                deckData.cards.length === 0) {
                throw new Error('No cards in API response');
            }
            setCardsLeft(response.data.remaining);

            const fetchedCard = deckData.cards[0];
            const { images, code, value, suit } = fetchedCard;

            const newCard = {
                png: images?.png || fetchedCard.image || '',
                code,
                value,
                suit,
                remaining: deckData.remaining,
                shown: true,
                rotationIndex: deck.length
            };

            setRotations([...rotations, generateRotation()]);
            setDeck(previousDeck => [...previousDeck, newCard]);
        } catch (error) {
            console.error('Error fetching card:', error);
        }
    };

    const showCard = () => {
        // console.log(deck);
        if (cardsLeft >= 0 &&
            cardsLeft <= deck.length) {
            deck[deck.length - cardsLeft].shown = true;
            setDeck([...deck]);
            setCardsLeft(prev => prev - 1);
        } else {
            console.warn("Invalid cardsLeft index:", cardsLeft);
        }
    };

    const getRemaining = () => cardsLeft;

    const remainingCards = getRemaining();

    const shuffledArray = prevDeck => {
        const uniqueCards = new Set(prevDeck);
        const shuffledDeck = Array.from(uniqueCards);
        for (let i = shuffledDeck.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [shuffledDeck[i], shuffledDeck[j]] =
                [shuffledDeck[j], shuffledDeck[i]];
        }
        shuffledDeck.forEach((card, index) => {
            card.shown = false;
            card.remaining = shuffledDeck.length - index - 1;
            card.rotationIndex = index;
        });
        setRotations(Array.from({ length: 52 }, () =>
            generateRotation()));
        setCardsLeft(52);
        return shuffledDeck;
    };

    const fetchNewDeck = async () => {
        setIsLoading(true);
        try {
            const response = await axios.get('https://deckofcardsapi.com/api/deck/new/shuffle/?deck_count=1');

            if (!response ||
                !response.data) {
                throw new Error('Invalid API response');
            }

            const deckData = response.data;

            if (!deckData ||
                !deckData.deck_id ||
                !deckData.remaining) {
                throw new Error('Unexpected API response format');
            }

            if (deckData.deck_id != null) {
                setDeckId(deckData.deck_id);
                setDeck([]);
                setCardsLeft(deckData.remaining);
            }
        } catch (error) {
            console.error('Error fetching new deck:', error.message);
            setDeckId(null);
            setDeck([]);
            setCardsLeft(-1);
        } finally {
            setIsLoading(false);
        }
    };

    const clearDeck = () => {
        setDeck([]);
        setDeckId(null);
        setCardsLeft(-1);
        setRotations([]);
    };

    const clearAndFetchNewDeck = () => {
        setDeck([]);
        setDeckId(null);
        setCardsLeft(-1);
        setRotations([]);
        fetchNewDeck();
    };

    const shuffleDeck = () => {
        if (deck.length === 0) {
            fetchNewDeck();
        } else {
            setDeckId(null);
            setDeck(prevDeck => shuffledArray(prevDeck));
        }
    }


    useEffect(() => {
        if (cardsLeft === 0) {
            alert("You are out of cards!");
        }
    }, [cardsLeft]);

    useEffect(() => {
        let isMounted = true;
        const fetchNewDeck = async () => {
            if (!isMounted) return;
            setIsLoading(true);
            clearDeck();

            try {
                const response = await axios.get('https://deckofcardsapi.com/api/deck/new/shuffle/?deck_count=1');
                const deckData = response.data;

                if (isMounted && deckData.deck_id != null) {
                    setDeckId(deckData.deck_id);
                    setDeck([]);
                    setCardsLeft(deckData.remaining);
                }
            } catch (error) {
                console.error('Error fetching new deck:', error);
            } finally {
                if (isMounted) {
                    setIsLoading(false);
                }
            }
        };

        fetchNewDeck();

        return () => {
            isMounted = false;
        };
    }, []);
    return (
        <>
            <h1>
                Card Flipper
            </h1>
            {isLoading ? (
                <div className="loading">Loading...</div>
            ) : (
                <>
                    <DeckControls
                        deckId={deckId}
                        deckState={deck}
                        shuffle={shuffleDeck}
                        clear={clearAndFetchNewDeck}
                        add={addCard}
                        show={showCard}
                        remaining={remainingCards}
                        toggleAutoDraw={toggleAutoDraw}
                        autoDrawEnabled={autoDrawEnabled}
                    />
                    <div className="cards">
                        {deck.map(currentCard => (
                            <Card
                                key={uuidv4()}
                                cardObject={currentCard}
                                rotation={rotations[currentCard.rotationIndex] || 'rotate(0deg)'}
                            />
                        ))}
                    </div>
                </>
            )}
        </>
    );
}

export default Deck;