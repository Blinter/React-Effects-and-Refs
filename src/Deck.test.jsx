// Deck.test.jsx

import React from 'react';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import axios from 'axios';
import Deck from './Deck';

jest.mock('axios');

describe('Deck component', () => {
    beforeEach(() => {
        jest.resetAllMocks();
    });

    test('fetches new deck on mount', async () => {
        axios.get.mockResolvedValueOnce({
            data: { deck_id: 'test_deck_id', remaining: 52 }
        });

        render(<Deck />);

        await waitFor(() => {
            expect(axios.get).toHaveBeenCalledWith('https://deckofcardsapi.com/api/deck/new/shuffle/?deck_count=1');
        });
    });

    test('displays remaining cards count', async () => {
        axios.get.mockResolvedValueOnce({
            data: { deck_id: 'test_deck_id', remaining: 52 }
        });

        render(<Deck />);

        await waitFor(() => {
            const remainingCount = screen.getByText(/Remaining: \d+/);
            expect(remainingCount).toBeInTheDocument();
            expect(remainingCount).toHaveTextContent('Remaining: 52');
        });
    });

    test('draws a card when Draw button is clicked', async () => {
        axios.get.mockResolvedValueOnce({
            data: { deck_id: 'test_deck_id', remaining: 52 }
        });

        axios.get.mockResolvedValueOnce({
            data: {
                cards: [{
                    png: 'image_url',
                    image: 'image_url',
                    value: 'ACE',
                    suit: 'SPADES',
                    code: 'AS'
                }],
                remaining: 51
            }
        });

        render(<Deck />);

        await waitFor(() => {
            expect(screen.getByText('Remaining: 52')).toBeInTheDocument();
        });

        const drawButton = screen.getByRole('button', { name: 'Draw a card' });
        expect(drawButton).toHaveAttribute('aria-label', 'Draw a card');
        fireEvent.click(drawButton);

        await waitFor(() => {
            expect(axios.get).toHaveBeenCalledWith('https://deckofcardsapi.com/api/deck/test_deck_id/draw/?count=1');
            expect(screen.getByText('Remaining: 51')).toBeInTheDocument();
        });
    });

    test('clears the deck when Clear button is clicked', async () => {
        axios.get.mockResolvedValueOnce({
            data: { deck_id: 'test_deck_id', remaining: 52 }
        });

        render(<Deck />);

        await waitFor(() => {
            expect(screen.getByText('Remaining: 52')).toBeInTheDocument();
        });

        const clearButton = screen.getByRole('button', { name: /clear/i });
        expect(clearButton).toHaveAttribute('aria-label', 'Clear the deck');
        fireEvent.click(clearButton);

        await waitFor(() => {
            expect(screen.getByText('Remaining: -1')).toBeInTheDocument();
        });
    });

    test('enables and disables auto-draw', async () => {
        axios.get.mockResolvedValueOnce({
            data: { deck_id: 'test_deck_id', remaining: 52 }
        });

        render(<Deck />);

        await waitFor(() => {
            expect(screen.getByText('Remaining: 52')).toBeInTheDocument();
        });

        const autoDrawButton = screen.getByRole('button', { name: /auto draw/i });
        expect(autoDrawButton).toHaveAttribute('aria-label', 'Start Auto Draw');
        expect(autoDrawButton).toHaveTextContent('Start Auto Draw');

        fireEvent.click(autoDrawButton);
        expect(autoDrawButton).toHaveTextContent('Stop Auto Draw');

        fireEvent.click(autoDrawButton);
        expect(autoDrawButton).toHaveTextContent('Start Auto Draw');
    });

    test('alerts when out of cards', async () => {
        axios.get.mockResolvedValueOnce({
            data: { deck_id: 'test_deck_id', remaining: 0 }
        });

        window.alert = jest.fn();

        render(<Deck />);

        await waitFor(() => {
            expect(window.alert).toHaveBeenCalledWith('You are out of cards!');
        });
    });

    test('matches initial snapshot', async () => {
        axios.get.mockResolvedValueOnce({
            data: { deck_id: 'test_deck_id', remaining: 52 }
        });

        const { container } = render(<Deck />);
        await waitFor(() => {
            expect(container).toMatchSnapshot();
        });
    });

    test('matches snapshot after drawing a card', async () => {
        axios.get.mockResolvedValueOnce({
            data: { deck_id: 'test_deck_id', remaining: 52 }
        });

        axios.get.mockResolvedValueOnce({
            data: {
                cards: [{
                    png: 'image_url',
                    image: 'image_url',
                    value: 'ACE',
                    suit: 'SPADES',
                    code: 'AS'
                }],
                remaining: 51
            }
        });

        const { container } = render(<Deck />);
        await waitFor(() => {
            expect(screen.getByText('Remaining: 52')).toBeInTheDocument();
        });

        const drawButton = screen.getByRole('button', { name: 'Draw a card' });
        fireEvent.click(drawButton);

        await waitFor(() => {
            expect(container).toMatchSnapshot();
        });
    });

    test('matches snapshot after clearing the deck', async () => {
        axios.get.mockResolvedValueOnce({
            data: { deck_id: 'test_deck_id', remaining: 52 }
        });

        const { container } = render(<Deck />);
        await waitFor(() => {
            expect(screen.getByText('Remaining: 52')).toBeInTheDocument();
        });

        const clearButton = screen.getByRole('button', { name: /clear/i });
        fireEvent.click(clearButton);

        await waitFor(() => {
            expect(container).toMatchSnapshot();
        });
    });

    test('matches snapshot when out of cards', async () => {
        axios.get.mockResolvedValueOnce({
            data: { deck_id: 'test_deck_id', remaining: 0 }
        });

        const { container } = render(<Deck />);
        await waitFor(() => {
            expect(container).toMatchSnapshot();
        });
    });

});
