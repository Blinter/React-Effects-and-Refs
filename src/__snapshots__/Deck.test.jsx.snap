// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deck component matches initial snapshot 1`] = `
<div>
  <h1>
    Card Flipper
  </h1>
  <div
    class="loading"
  >
    Loading...
  </div>
</div>
`;

exports[`Deck component matches snapshot after clearing the deck 1`] = `
<div>
  <h1>
    Card Flipper
  </h1>
  <div
    class="loading"
  >
    Loading...
  </div>
</div>
`;

exports[`Deck component matches snapshot after drawing a card 1`] = `
<div>
  <h1>
    Card Flipper
  </h1>
  <div
    class="deck-controls"
  >
    <button
      aria-label="Draw a card"
    >
      Draw
    </button>
    |
    <button
      aria-label="Clear the deck"
    >
      Clear
    </button>
    |
    <button
      aria-label="Start Auto Draw"
    >
      Start Auto Draw
    </button>
    <p
      class="remaining-cards"
    >
      Remaining: 
      52
    </p>
  </div>
  <div
    class="cards"
  />
</div>
`;

exports[`Deck component matches snapshot when out of cards 1`] = `
<div>
  <h1>
    Card Flipper
  </h1>
  <div
    class="loading"
  >
    Loading...
  </div>
</div>
`;
